feat: allow repl pause on exception state to be set from env

iloveitaly opened the pull request:
Right now, it's not possible to configure the debugger to pause on uncaught exceptions
without manually configuring the pause state on each run. This change would allow the
pause state to be configured via a shell env variable to allow for faster CLI-based node
debugging.

I'm not sure if there's a reason there aren't more configuration options for `node inspect`,
so I wanted to submit this PR for discussion before writing tests and updating documentation.

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

iloveitaly committed:
feat: allow repl pause on exception state to be set from env

Right now, it's not possible to configure the debugger to pause on uncaught exceptions
without manually configuring the pause state on each run. This change would allow the
pause state to be configured via a shell env variable to allow for faster CLI-based node
debugging.

ryanstout commented:
+1 thanks!

Trott commented:
I know there's usually resistance to adding environment variables even moreso than CLI flags, but I'm not sure there's a better route to enable this feature....

@nodejs/post-mortem (Is that even the right group to ping for the CLI debugger?)

aduh95 commented:
We would want the env variable to be documented along side the other env variables Node.js supports.

Trott commented:
> We would want the env variable to be documented along side the other env variables Node.js supports.

True. The OP acknowledged that:

> I'm not sure if there's a reason there aren't more configuration options for node inspect,
so I wanted to submit this PR for discussion before writing tests and updating documentation.

Maybe this should be in draft mode.

Anyway, I'm trying to remember who has expressed something at least vaguely like "please, no more environment variables" opinions in the past. Maybe it was @cjihrig?

cjihrig commented:
> Anyway, I'm trying to remember who has expressed something at least vaguely like "please, no more environment variables" opinions in the past. Maybe it was @cjihrig?

I don't think it was me - at least not originally. IIRC, the agreed upon way to move forward recently has been adding CLI flags and passing those via the `NODE_OPTIONS` environment variable.

GeoffreyBooth commented:
Since you can always set flags via the `NODE_OPTIONS` variable, like `NODE_OPTIONS=--inspect`, adding flags is considered preferable as then Node can be configured via _either_ flags or an environment variable. Can this PR be refactored to use a flag instead?

If the new option applies _only_ to `node inspect`, it could be a flag that comes after that like `node inspect --pause-on-uncaught-exceptions` or whatever, where this flag exists for `node inspect` but not for `node` generally. This wouldn’t be settable via environment variable, though, but since `node inspect` is interactive I’m not sure its options _need_ to be configured via environment variable?

Trott commented:
Over in a Slack private conversation, @GeoffreyBooth suggested that maybe there can be `inspect`-specific flags. `node inspect --pause-on-exception foo.js` and things like that.  That might present challenges for the `NODE_OPTIONS` part of it, though. But it might make sense if we think there are going to be a lot of these.

And on the other hand, there already is the `NODE_INSPECT_RESUME_ON_START` environment variable, so that could argue for continuing down that path, I guess.

Since this is for a developer CLI convenience, I think we'd want something that they could set once and never have to think about again. That implies an environment variable, and it also implies one that *isn't* `NODE_OPTIONS` since we wouldn't want to include things that could interfere with normal Node.js operations, just CLI debugging.

I should probably collect my thoughts and come up with a recommendation rather than all this listing out of pros and cons.....

bnoordhuis commented:
> And on the other hand, there already is the NODE_INSPECT_RESUME_ON_START environment variable, so that could argue for continuing down that path, I guess.

To counter-argue: that was introduced when node-inspect was moved into core, i.e., a third-party indiscretion.

FWIW, I was probably the one who first pushed back against adding yet more environment variables. They're essentially ambient state and that's something to avoid as much as possible.

Trott commented:
> FWIW, I was probably the one who first pushed back against adding yet more environment variables. They're essentially ambient state and that's something to avoid as much as possible.

Works for me. I'm happy to be on board with "use CLI flags and set them in `NODE_OPTIONS`" as a general rule.

Honestly, if we were designing this particularly thing from scratch, I wonder if the approach we'd want is to read some kind of config file.

GeoffreyBooth commented:
> What I'd like to propose is in addition to the above is:
> 
> 1. Initially, add support for a `NODE_INSPECT_OPTIONS` env var which could be used to replace the bespoke `NODE_INSPECT_RESUME_ON_START` var (and any other similar vars if they exist)
> 2. In the future, a `~/.nodeinspectrc` file, with the name, format (json/javascript/both), and available options to be debated at a later time.

I like the first one. I think ideally, we would add support for `node inspect`-only flags like `node inspect --pause-on-exception-state` where this flag `--pause-on-exception-state` works for `node inspect` but not for `node`. Then yes, we also create `NODE_INSPECT_OPTIONS` where it can follow the pattern of `NODE_OPTIONS` so like `NODE_INSPECT_OPTIONS=--pause-on-exception-state`.

Re `~/.nodeinspectrc`, there are already a few open issues about supporting configuring Node via files. I would prefer to adopt a general solution for supporting _all_ of Node’s configuration options via a file, rather than a config file that works only for `node inspect`. One approach already discussed would be to give Node [support for `.env` files](https://github.com/nodejs/node/pull/43973#issuecomment-1249549346) like Bun has; that allows for both configuring Node (if `NODE_OPTIONS` and/or `NODE_INSPECT_OPTIONS` are defined in the file) as well as setting whatever other environment variables the user would want defined for the application. Besides being very powerful, this approach seems likely to be the easiest solution to implement, as parsing a `.env` file within the C++ parser is probably more straightforward than JSON or running JavaScript, and we can always add support for JSON or other formats later in addition.

iloveitaly committed:
feat: node inspect argv parser, allow exception state to be set from command line

iloveitaly commented:
Update here for @GeoffreyBooth, @Trott, and team!

- I've added support for `NODE_INSPECT_OPTIONS` although I've not added support for `--port` and other options there. Curious if you think this is needed right away, or we can do this in a separate PR. I'm wary of making breaking changes (even small ones)
- Added support for `--pause-on-exception-state` and `--inspect-resume-on-start`

This is working great on my [`better-node-inspect`](https://github.com/iloveitaly/better-node-inspect) library, but there's still work left to do here:

- Add tests, if applicable (any points here?)
- Run more manual tests

Wanted to post here and get some initial feedback on the approach before I go farther.

Trott commented:
> * Add tests, if applicable (any points here?)

Existing tests for the CLI debugger are in `test/parallel/test-debugger*` and (for tests that use an explicit port number or otherwise can't be run reliably while other tests are running) `test/sequential/test-debugger*`. 

The helper library for CLI debugger tests is documented in `test/common/README.md` which can also be accessed at https://github.com/nodejs/node/tree/main/test/common#debugger-module. (Looking at it now, the debugger helper stuff can use considerably more documentation. But hopefully the function names are clear enough along with the existing tests for reference.)

For non-debugger-specific stuff about tests, there is a guide at in `doc/contributing/writing-tests.md` which can also be accessed at https://github.com/nodejs/node/blob/main/doc/contributing/writing-tests.md.

You can also ask questions about writing tests for Node.js core in the `#nodejs-core` channel on the OpenJS Slack: https://openjs-foundation.slack.com/archives/C019Y2T6STH

Trott commented:
> Existing tests for the CLI debugger are in `test/parallel/test-debugger*` and (for tests that use an explicit port number or otherwise can't be run reliably while other tests are running) `test/sequential/test-debugger*`.

Oh, it appears that there are also a small number of them in `test/parallel/test-inspect*`. Unless there's a reason for the discrepancy, it would probably be a good idea to get them all the CLI debugger tests to be named consistently as either `test-debugger*` or `test-inspect*`.
